### PR TITLE
[lldb/Interpreter] Introduce `ScriptedStopHook{,Python}Interface` & make use of it

### PR DESCRIPTION
**Pull request by medismailben**
This patch introduces new `ScriptedStopHook{,Python}Interface` classes that make use of the Scripted Interface infrastructure and makes use of it in `StopHookScripted`.

It also relax the requirement on the number of argument for initializing scripting extension if the size of the interface parameter pack contains 1 less element than the extension maximum number of positional arguments for this initializer.
This addresses the cases where the embedded interpreter session dictionary is passed to the extension initializer which is not used most of the time.

### CODE REVIEW

File: lldb/source/Plugins/ScriptInterpreter/Python/Interfaces/ScriptedPythonInterface.h [hunk 1]
```diff
@@ -137,12 +137,28 @@ class ScriptedPythonInterface : virtual public ScriptedInterface {
...
+            },
+            std::tuple_cat(transformed_args, std::make_tuple(dict)));
+      } else {
+        std::apply(
+            [&init, &expected_return_object](auto &&...args) {
+              llvm::consumeError(expected_return_object.takeError());
+              expected_return_object = init(args...);
+            },
+            transformed_args);
+      }
```
Comment (JDevlieghere) [inline lldb/source/Plugins/ScriptInterpreter/Python/Interfaces/ScriptedPythonInterface.h:161]: Your commit message explains it, but I think this deserves a comment here as well.

**Review by JDevlieghere — APPROVED** [review]
LGTM!

**Review by bulbazord — APPROVED** [review]

**Comment by medismailben** [comment]
Depends on https://github.com/llvm/llvm-project/pull/109063/files

File: lldb/unittests/ScriptInterpreter/Python/PythonTestSuite.cpp [hunk 2]
```diff
@@ -269,21 +274,7 @@ void *lldb_private::python::SWIGBridge::LLDBSWIGPython_GetDynamicSetting(
...
-}
-
-bool lldb_private::python::SWIGBridge::LLDBSwigPythonStopHookCallHandleStop(
-    void *implementor, lldb::ExecutionContextRefSP exc_ctx_sp,
-    lldb::StreamSP stream) {
-  return false;
-}
-
-python::PythonObject
-lldb_private::python::SWIGBridge::ToSWIGWrapper(Status status) {
```
Comment (DavidSpickett) [inline lldb/unittests/ScriptInterpreter/Python/PythonTestSuite.cpp:286]: FYI https://github.com/llvm/llvm-project/commit/76b827bb4d5b4cc4d3229c4c6de2529e8b156810

**Comment by DavidSpickett** [comment]
I've reverted this (https://github.com/llvm/llvm-project/commit/801046e3303eed43bffebb84e9e505cc19cad5c0) because it broke a stop hook test:
https://lab.llvm.org/buildbot/#/builders/162/builds/6735

It seems to add the stop hook but there's no output when it should have been called.

**Pull request merged**